pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -769,59 +791,64 @@ func (pl *dynamicResources) PreScore(ctx context.Context, cs *framework.CycleSta
+		return nil
+	}
+	if haveAllPotentialNodes(state.podSchedulingState.schedulingCtx, nodes) {
+		logger.V(5).Info("all potential nodes already set", "pod", klog.KObj(pod), "potentialnodes", klog.KObjSlice(nodes))
+		return nil
+	}
+
+	// Remember the potential nodes. The object will get created or
+	// updated in Reserve. This is both an optimization and
+	// covers the case that PreScore doesn't get called when there
+	// is only a single node.
+	logger.V(5).Info("remembering potential nodes", "pod", klog.KObj(pod), "potentialnodes", klog.KObjSlice(nodes))
+	numNodes := len(nodes)
+	if numNodes > resourcev1alpha2.PodSchedulingNodeListMaxSize {
+		numNodes = resourcev1alpha2.PodSchedulingNodeListMaxSize
+	}
+	potentialNodes := make([]string, 0, numNodes)
+	if numNodes == len(nodes) {
+		// Copy all node names.
+		for _, node := range nodes {
+			potentialNodes = append(potentialNodes, node.Name)
 		}
-		schedulingCtx.Spec.PotentialNodes = make([]string, 0, numNodes)
-		if numNodes == len(nodes) {
-			// Copy all node names.
-			for _, node := range nodes {
-				schedulingCtx.Spec.PotentialNodes = append(schedulingCtx.Spec.PotentialNodes, node.Name)
-			}
-		} else {
-			// Select a random subset of the nodes to comply with
-			// the PotentialNodes length limit. Randomization is
-			// done for us by Go which iterates over map entries
-			// randomly.
-			nodeNames := map[string]struct{}{}
-			for _, node := range nodes {
-				nodeNames[node.Name] = struct{}{}
-			}
-			for nodeName := range nodeNames {
-				if len(schedulingCtx.Spec.PotentialNodes) >= resourcev1alpha2.PodSchedulingNodeListMaxSize {
-					break
-				}
-				schedulingCtx.Spec.PotentialNodes = append(schedulingCtx.Spec.PotentialNodes, nodeName)
+	} else {
+		// Select a random subset of the nodes to comply with
+		// the PotentialNodes length limit. Randomization is
+		// done for us by Go which iterates over map entries
+		// randomly.
+		nodeNames := map[string]struct{}{}
+		for _, node := range nodes {
+			nodeNames[node.Name] = struct{}{}
+		}
+		for nodeName := range nodeNames {
+			if len(potentialNodes) >= resourcev1alpha2.PodSchedulingNodeListMaxSize {
+				break
 			}
+			potentialNodes = append(potentialNodes, nodeName)
 		}
-		sort.Strings(schedulingCtx.Spec.PotentialNodes)
-		state.storePodSchedulingContexts(schedulingCtx)
 	}
-	logger.V(5).Info("all potential nodes already set", "pod", klog.KObj(pod), "potentialnodes", klog.KObjSlice(nodes))
+	sort.Strings(potentialNodes)
+	state.podSchedulingState.potentialNodes = &potentialNodes
+	state.preScored = true
 	return nil
 }
 
-func haveAllNodes(nodeNames []string, nodes []*v1.Node) bool {
+func haveAllPotentialNodes(schedulingCtx *resourcev1alpha2.PodSchedulingContext, nodes []*v1.Node) bool {
+	if schedulingCtx == nil {
+		return false
+	}
 	for _, node := range nodes {
-		if !haveNode(nodeNames, node.Name) {
+		if !haveNode(schedulingCtx.Spec.PotentialNodes, node.Name) {
```

Comment:
This is `O(N^2)`. What was the limit for PotentialNodes? I think nodes could be all nodes in a cluster, so maybe it's worth using a `Set`

Reply:
The size limit is 128. I've looked at CPU profiles for scheduler_perf runs with 100 nodes and none of the "is a node name listed" checks show up. I don't see a reason to make the code more complex.

Reply:
FWIW, I also experimented with directly using a `Set` in the API. My reasoning was that if most of the code has a set readily available and the Set only needs to be built when decoding, then perhaps performance will become better - it became worse, because encoding/decoding a Set was more expensive than iterating through slices.

Reply:
> The size limit is 128. I've looked at CPU profiles for scheduler_perf runs with 100 nodes and none of the "is a node name listed" checks show up. 

What about 5k nodes, the official supported max number of nodes.

Reply:
> I also experimented with directly using a Set in the API. 

I don't think we do that anywhere in the API. But building a Set out of a slice seems reasonable. In any case, probably worth exploring separately.

Reply:
> What about 5k nodes, the official supported max number of nodes.

The slices will get truncated. I've simulated such larger clusters with scheduler_perf and it still worked.

Reply:
About performance: I remember that using a set had been investigated already in https://github.com/kubernetes/kubernetes/pull/115930

The performance benefit wasn't obvious. The conclusion was to require instead that these slices must be sorted: https://github.com/kubernetes/kubernetes/issues/116416

That simplifies various pieces, ranging from validation to these searches.

It's an API break, so has to wait for the actual alpha -> beta API change. But I can prototype it now... on it.

Reply:
Are you suggesting to change the API to use set directly? Before you waste any time, worth asking API reviewers if they would accept such change. My guess would be no.

Reply:
FWIIW, I don't see any guidelines against it https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md, except maybe for https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

Reply:
> Are you suggesting to change the API to use set directly? Before you waste any time, worth asking API reviewers if they would accept such a change. My guess would be no.

I'd rather waste my time than the one from API reviewers :sweat_smile:  So I tried it first to see whether there is any performance improvement and how dirty the solution would be - no benefit and quite hacky, so I won't pursue that angle further.

Reply:
What I am probably going to propose instead is to require that the slices must be sorted, but let's discuss that when I have a PR ready.

Reply:
SGTM. Make sure to add it to API validation. And that the list is `atomic`, if it wasn't already.